### ISSUE TITLE
Pyright possibly confuses constrained type specification with union type

### ISSUE THREAD

**Issue by rk-terence** [microsoft/pyright#6521]
**Describe the bug**
I think when working with type parameters that have constrained type specifications, we may encounter a bug of Pyright, where Pyright possibly confuses the constrained type specification with the union type.
It is shown by the code snippet below.

**Code or Screenshots**
```python
from typing import Iterable


def f(iterable: Iterable[str] | Iterable[int]) -> None:
    for ele in iterable:
        print(ele)


type StrOrIntegerIterable[T: (str, int)] = Iterable[T]


def g[T: (str, int)](iterable: StrOrIntegerIterable[T]):
    f(iterable)  # this line fails type checking with pyright


def g1(iterable: StrOrIntegerIterable[str]):
    f(iterable)  # this line passes type checking with pyright


def g2(iterable: StrOrIntegerIterable[int]):
    f(iterable)  # this line passes type checking with pyright
```

The error reported by running Pyright CLI is
```
  /scratch.py:13:7 - error: Argument of type "StrOrIntegerIterable[T@g]" cannot be assigned to parameter "iterable" of type "Iterable[str] | Iterable[int]" in function "f"
    Type "StrOrIntegerIterable[T@g]" cannot be assigned to type "Iterable[str] | Iterable[int]"
      "StrOrIntegerIterable[T@g]" is incompatible with "Iterable[str]"
        Type parameter "_T_co@Iterable" is covariant, but "T@g" is not a subtype of "str"
          Type "str* | int*" cannot be assigned to type "str"
            "int*" is incompatible with "str"
      "StrOrIntegerIterable[T@g]" is incompatible with "Iterable[int]"
        Type parameter "_T_co@Iterable" is covariant, but "T@g" is not a subtype of "int"
          Type "str* | int*" cannot be assigned to type "int"
    ... (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```

AFAIK, since the type parameter `T` in `g` is constrained to `int` or `str`, the argument `iterable` in function `g` is either `Iterable[str]` or `Iterable[int]`, which meets the requirements of the signature of `f`.
My understanding is that "`T` is either a subtype of A or a subtype of B" is different from "`T` is a subtype of Union[A, B]", which can be verified using
```python
def h[T: (str, int)](x: T) -> T:
    return x

def caller(a: int | str) -> None:
    h(a)  # this line fails type checking with pyright
```

So I think my code snippet above should pass the type check.
Is this a bug of Pyright, or have I missed some important concepts?

**VS Code extension or command-line**
pyright CLI version 1.1.336

**Comment by rk-terence** [comment]
I encountered this when when working with `np.floating[Any]` and `float`.
Both `Sequence[np.floating[Any]]` and `Sequence[float]` are subtypes of `numpy.typing.ArrayLike`, but when I want my function to accept both `Sequence[np.floating[Any]]` and `Sequence[float]` using type constraints syntax, the type checking failed.

Please check the below snippet.

```python
from typing import Sequence, Any

import numpy as np
from numpy.typing import ArrayLike


def func(x: ArrayLike) -> None:
    pass


def caller[T: (np.floating[Any], float)](x: Sequence[T]):
    func(x)  # this fails type checking


def caller1(x: Sequence[np.floating[Any]]):
    func(x)  # this passes


def caller2(x: Sequence[float]):
    func(x)  # this passes
```

**Comment by erictraut** [comment]
I don't consider this a bug. It's a limitation given the way pyright implements constrained TypeVars. See [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=constrained-type-variables) for details. This happens to work in mypy given the way it implements constrained TypeVars, but mypy's approach doesn't work in other areas where pyright does.

**Comment by rk-terence** [comment]
Thank you for your reply.

I think I may use a workaround here by annotating the `caller` function's `x` as `Sequence[np.floating[Any]] | Sequence[float]`.
